### PR TITLE
CI Refactor - Top level Rakefile

### PR DESCRIPTION
**Pull request by alextwoods**
*Description of changes:*

Follows the pattern from: https://github.com/smithy-lang/smithy-ruby/pull/179

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .github/workflows/gems_ci.yml [hunk 1]
```diff
@@ -25,28 +29,67 @@ jobs:
...
+      - name: Set up JDK
+        uses: actions/setup-java@v1
+        with:
+          java-version: ${{ env.java_version }}
+
+      - name: Build and publish maven local smithy-ruby
+        run: |
+          git clone --depth 1 https://github.com/smithy-lang/smithy-ruby.git
+          cd smithy-ruby/codegen
+          ./gradlew clean build publishToMavenLocal
```
Comment (mullermp) [inline .github/workflows/gems_ci.yml:50]: Any way we can load rakefiles from smithy ruby instead? and have it run the build/publish rake tasks?

File: .github/workflows/protocol_test_ci.yml [hunk 2]
```diff
@@ -108,22 +29,36 @@ jobs:
...
+          java-version: ${{ env.java_version }}
 
-      - name: Copy Gemfile
+      - name: Build and publish maven local smithy-ruby
         run: |
-          cp Gemfile ${{ matrix.protocol }}
+          git clone --depth 1 https://github.com/smithy-lang/smithy-ruby.git
+          cd smithy-ruby/codegen
+          ./gradlew clean build publishToMavenLocal
+          cd ../..
```
Comment (mullermp) [inline .github/workflows/protocol_test_ci.yml:51]: Instead of cd back and forth, can we just set a working dir here?

File: Gemfile [hunk 3]
```diff
@@ -18,9 +20,15 @@ end
 group :development do
   gem 'rbs'
   gem 'rubocop'
+  gem 'racc'
```
Comment (mullermp) [inline Gemfile:23]: What's this for?

File: Rakefile [hunk 4]
```diff
@@ -1,9 +1,123 @@
 # frozen_string_literal: true
```
Comment (mullermp) [inline Rakefile:1]: Any notable differences between this and smithy-ruby's rakefile?

File: codegen/smithy-aws-ruby-codegen/src/main/resources/ruby/customizations/s3/endpoint_rules.rb [hunk 5]
```diff
@@ -1,5 +1,7 @@
+# S3 Endpoint custom functions
 module Endpoints
   # aws.isVirtualHostableS3Bucket(value: string, allowSubDomains: bool) bool
+  # rubocop:disable Style/OptionalBooleanParameter
```
Comment (mullermp) [inline codegen/smithy-aws-ruby-codegen/src/main/resources/ruby/customizations/s3/endpoint_rules.rb:4]: Would be nice to fix this one .... I tried

File: gems/aws-sigv4/spec/spec_helper.rb [hunk 6]
```diff
@@ -1,11 +1,6 @@
 # frozen_string_literal: true
 
 require 'rspec'
-require 'simplecov'
```
Comment (mullermp) [inline gems/aws-sigv4/spec/spec_helper.rb:4]: What happened to simplecov? I don't think sigv4 should share from core if that's what's happening. This is stand alone.

File: gems/aws-sigv4/lib/aws-sigv4/signer.rb [hunk 7]
```diff
@@ -79,13 +79,15 @@ module SigV4
     #   )
     #
     class Signer
+      # rubocop:disable Style/ClassVars
```
Comment (mullermp) [inline gems/aws-sigv4/lib/aws-sigv4/signer.rb:82]: Is there a better solution here that rubocop recommends?

**Review by mullermp — COMMENTED** [review]
Looks good so far!

File: codegen/smithy-aws-ruby-codegen/src/main/resources/ruby/customizations/s3/endpoint_rules.rb (hunk 5 above)
Comment (alextwoods) [reply codegen/smithy-aws-ruby-codegen/src/main/resources/ruby/customizations/s3/endpoint_rules.rb]: Fixed - just removed the default. This matches the endpoints language definitions of the function more closely anyway (which ofc does not define/describe default function values).

File: gems/aws-sigv4/spec/spec_helper.rb (hunk 6 above)
Comment (alextwoods) [reply gems/aws-sigv4/spec/spec_helper.rb:4]: It gets started twice because this file requires the core spec_helper.  

The coverage is still standalone.

Comment (alextwoods) [reply gems/aws-sigv4/spec/spec_helper.rb:4]: This is a consequenece of running the tests together.  We can only start simple cov once.

Alternatives - create seperate targets for core and sigv4 and never run the tests at the same time.

File: .github/workflows/gems_ci.yml (hunk 1 above)
Comment (alextwoods) [reply .github/workflows/gems_ci.yml:50]: Added task here: https://github.com/smithy-lang/smithy-ruby/commit/2cfc306c1dea68ceb8550cbffd7187613c3a5fda

And will use.

File: .github/workflows/protocol_test_ci.yml (hunk 2 above)
Comment (alextwoods) [reply .github/workflows/protocol_test_ci.yml]: no - the working directory doesn't exist before this

File: Gemfile (hunk 3 above)
Comment (alextwoods) [reply Gemfile:23]: required for rbs to work here and CI to pass.  Maybe not required anymore (may have been a temporarily missing dependncy in rbs?)

Comment (alextwoods) [reply Gemfile:23]: I get this error without it:
```
Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.
/Users/alexwoo/.gem/ruby/3.3.0/gems/parser-3.1.2.0/lib/parser.rb:12:in `require': cannot load such file -- racc/parser (LoadError)
```

Comment (alextwoods) [reply Gemfile:23]: Edit: racc is a bundled gem in 3.3, so needs to be included going forward: https://stdgems.org

File: Rakefile (hunk 4 above)
Comment (alextwoods) [reply Rakefile:1]: The general structure is the same, but the tasks are fairly different.

File: gems/aws-sigv4/lib/aws-sigv4/signer.rb (hunk 7 above)
Comment (alextwoods) [reply gems/aws-sigv4/lib/aws-sigv4/signer.rb:82]: You can use a class instance variable instead. The behavior is different and implications for inheritance.

In our case, we want a class var and not a class instance variable I believe.

File: gems/aws-sigv4/spec/spec_helper.rb (hunk 6 above)
Comment (mullermp) [reply gems/aws-sigv4/spec/spec_helper.rb:4]: Yeah I think we should. It's a separate gem, which core will depend on.

**Review by mullermp — APPROVED** [review]
Approving but fix simplecov I think

**Pull request merged**